explorer/interpreter/type_checker.cpp

```diff
@@ -4232,6 +4222,49 @@ auto TypeChecker::DeclareInterfaceDeclaration(
   for (Nonnull<Declaration*> m : iface_decl->members()) {
     CARBON_RETURN_IF_ERROR(DeclareDeclaration(m, iface_scope_info));
 
+    // For an `extends C;` declaration, add `Self is C` to our constraint.
+    if (auto* extends = dyn_cast<InterfaceExtendsDeclaration>(m)) {
+      CARBON_ASSIGN_OR_RETURN(Nonnull<const Value*> base,
+                              TypeCheckTypeExp(extends->base(), iface_scope));
+      auto* constraint_type = dyn_cast<ConstraintType>(base);
+      if (auto* interface_type = dyn_cast<InterfaceType>(base)) {
+        CARBON_ASSIGN_OR_RETURN(
+            constraint_type,
+            MakeConstraintForInterface(extends->source_loc(), interface_type));
+      }
+      if (!constraint_type) {
+        return ProgramError(extends->source_loc())
+               << "an interface can only extend a constraint, found " << *base;
+      }
+      CARBON_RETURN_IF_ERROR(builder.AddAndSubstitute(
+          *this, constraint_type, builder.GetSelfType(),
+          builder.GetSelfWitness(), Bindings(),
+          /*add_lookup_contexts=*/true));
```

Comment:
This and lines 4252-4265 seem very structurally similar. Does it seem worth sharing to you?

Reply:
They're similar but the details end up being just different enough that I'm not sure it's worth it. I think it would be worth factoring out the code to convert a type-of-type to a `ConstraintType`; that comes up in a bunch of places, not only here. I'll do that in a separate PR.